### PR TITLE
Fix Kafka polling method

### PR DESCRIPTION
**Pull request by Arkatufus**
On previous polling implementation, when an exception happened during a poll, all of the messages consumed before the exception was thrown were discarded. The last offset/message pushed into the stream can be different, depending on how fast the broker and client exchanges messages and when the polling timer ran.

### CODE REVIEW

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 1]
```diff
@@ -545,22 +537,45 @@ private void Poll()
...
+        }
+
+        private (List<ConsumeResult<K, V>>, Exception) PollKafka(CancellationToken token)
         {
             ConsumeResult<K, V> consumed = null;
             var i = 10; // 10 poll attempts
             var timeout = Math.Max((int) _pollTimeout.TotalMilliseconds / i, 1);
-            var pooled = new List<ConsumeResult<K, V>>();
+            var polled = new List<ConsumeResult<K, V>>();
             do
```
Comment (Arkatufus) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:562]: New polling method. Instead of discarding all the consumed message when an exception is thrown, we return both the polled messages before the exception and the exception. We then first push all of the consumed messages and then process the exception. This makes polling consistent on both slow and fast brokers and allows us to implement correct stage restart behavior.

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 2]
```diff
@@ -475,67 +477,57 @@ private void Poll()
...
+                var resumeThese = currentAssignment.Where(partitionsToFetch.Contains).ToList();
+                var pauseThese = currentAssignment.Except(resumeThese).ToList();
+                PausePartitions(pauseThese);
+                ResumePartitions(resumeThese);
+
+                using (var cts = new CancellationTokenSource(_settings.PollTimeout))
+                {
+                    var (polled, exception) = PollKafka(cts.Token);
+                    ProcessResult(partitionsToFetch, polled);
+                    ProcessExceptions(exception);
```
Comment (Arkatufus) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:527]: Process all messages that happened before we hit any exceptions, then process the exception, if any.

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 3]
```diff
@@ -545,22 +537,45 @@ private void Poll()
             }
         }
 
-        private List<ConsumeResult<K, V>> PollKafka(CancellationToken token)
+        private void ProcessExceptions(Exception exception)
```
Comment (Arkatufus) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:540]: All exceptions are routed here, we don't kill the actor on all errors anymore, we keep it alive if the Decider returns Directive.Resume.

Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:540]: The side effect of this is the Decider delegate gets called twice, inside the KafkaConsumerActor and inside the stage itself.

**Comment by Aaronontheweb** [comment]
Wow, looks like a big change! I know we had discussed this but had no idea you were getting to work on it right away. Good stuff!

**Comment by Arkatufus** [comment]
I'm wondering if thrown exceptions should be emitted on the next polling cycle instead of right away, to let the stream catch up with the messages, if it were to fail instead of resume/restart.

**Comment by Arkatufus** [comment]
@Aaronontheweb Yeah, this is required to fix one of the racy specs, if not both.

**Comment by Arkatufus** [comment]
Plus I need to update some of the specs since we changed the default stream behavior, rechecked that they should be doing what they were supposed to be doing

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 4]
```diff
@@ -323,14 +332,7 @@ private void ApplySettings(ConsumerSettings<K, V> updatedSettings)
 
                 var localSelf = Self;
                 _consumer = _settings.CreateKafkaConsumer(
-                    consumeErrorHandler: (c, e) =>
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:326]: Interesting - so we just deliver a message to ourselves and handle it inside the actor's scope rather than externally now? Seems like a good idea, but why this change? Is it just to make sure that the decider is "in charge?"

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:540]: Only gets called twice in the event that the actor propagates the exception, right?

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 5]
```diff
@@ -545,22 +537,45 @@ private void Poll()
...
+                catch (Exception e)
+                {
+                    return (polled, e);
+                }
                 if (consumed != null)
-                    pooled.Add(consumed);
+                    polled.Add(consumed);
                 i--;
-            } while (i > 0 && !token.IsCancellationRequested);
+            } while (i > 0 && consumed != null && !token.IsCancellationRequested);
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:576]: So this will exit right away as soon as we receive a single message - no point in building a list of them if that's the case, no?

**Review by Aaronontheweb — COMMENTED** [review]
Left some questions

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs (hunk 5 above)
Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:576]: No, Consumer.Consume will return a null if its internal message queue is empty. So instead of waiting for the whole timeout duration, we just bail out of the polling method and return the messages that we polled up to that point.

Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:576]: You can consider the Consumer.Consume returning null the same as C string null terminator

File: src/Akka.Streams.Kafka/Stages/Consumers/Abstract/KafkaSourceStage.cs [hunk 6]
```diff
@@ -38,10 +38,12 @@ protected KafkaSourceStage(string stageName)
         {
             StageName = stageName;
             Shape = new SourceShape<TMessage>(Out);
+            InitialAttributes = Attributes.CreateName(StageName)
+                .And(ActorAttributes.CreateSupervisionStrategy(new DefaultConsumerDecider(true).Decide));
         }
 
         /// <inheritdoc />
-        protected override Attributes InitialAttributes => Attributes.CreateName(StageName);
+        protected override Attributes InitialAttributes { get; }
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Abstract/KafkaSourceStage.cs:46]: I was thinking more of a single instantiation, instead of having the attributes getting instantiated every time get_InitialAttributes is called.

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs (hunk 4 above)
Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:326]: Yes, the consumeErrorHandler callback is called on internal errors like unknown topic or partition. Not all errors are propagated to the `Consume` method. By routing this error to the actor, error handling become a lot more consistent, Plus we avoid any possible threading problems.

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 7]
```diff
@@ -475,67 +477,57 @@ private void Poll()
...
+                ResumePartitions(resumeThese);
+
+                using (var cts = new CancellationTokenSource(_settings.PollTimeout))
+                {
+                    var (polled, exception) = PollKafka(cts.Token);
+                    ProcessResult(partitionsToFetch, polled);
+                    ProcessExceptions(exception);
+                }
             }
-            // Workaroud for https://github.com/confluentinc/confluent-kafka-dotnet/issues/1366
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:522]: This is handled inside the decider now, that way it can actually be overridden if the user wishes to.

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs (hunk 3 above)
Comment (Arkatufus) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:540]: For now, it is always processed twice. It can be changed so that it doesn't propagate Resume directived errors.
I always propagate it because logging is handled by the stage, I've tried to make the actor as dumb as possible and hand over all responsibilities to the stage.

File: src/Akka.Streams.Kafka.Tests/BugFix240SupervisionStrategy.cs [hunk 8]
```diff
@@ -162,20 +162,27 @@ await Source.From(Enumerable.Range(1, 10))
...
             {
-                probe.ExpectNext(i, TimeSpan.FromSeconds(10)); 
+                var msg = probe.ExpectNext();
+                pulled.Add(msg);
             }
+
+            probe.ExpectNoMsg(TimeSpan.FromSeconds(2));
             probe.Cancel();
+
+            pulled.Should().BeEquivalentTo(new[] {1, 2, 3, 4, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
```
[the remark this answers is not in the document]
Comment (Arkatufus) [reply src/Akka.Streams.Kafka.Tests/BugFix240SupervisionStrategy.cs:181]: That... I'm not sure off, I'll come back to you.

File: src/Akka.Streams.Kafka.Tests/BugFix240SupervisionStrategy.cs [hunk 9]
```diff
@@ -162,20 +162,27 @@ await Source.From(Enumerable.Range(1, 10))
...
             {
-                probe.ExpectNext(i, TimeSpan.FromSeconds(10)); 
+                var msg = probe.ExpectNext();
+                pulled.Add(msg);
             }
+
+            probe.ExpectNoMsg(TimeSpan.FromSeconds(2));
             probe.Cancel();
+
+            pulled.Should().BeEquivalentTo(new[] {1, 2, 3, 4, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, opt => opt.WithStrictOrdering());
```
Comment (Arkatufus) [inline src/Akka.Streams.Kafka.Tests/BugFix240SupervisionStrategy.cs:181]: This now checks for strict ordering

File: src/Akka.Streams.Kafka/Stages/Consumers/Abstract/KafkaSourceStage.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/Akka.Streams.Kafka/Stages/Consumers/Abstract/KafkaSourceStage.cs:46]: Ah ok

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs (hunk 5 above)
Comment (Aaronontheweb) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:576]: Ok got it - I read the boolean logic backwards there

**Review by Aaronontheweb — APPROVED** [review]
LGTM - can merge when out of draft

**Comment by Aaronontheweb** [comment]
All of the partitioned source specs are failing @Arkatufus
